rcl/test/rcl/test_rmw_impl_id_check_func.cpp

```diff
@@ -70,3 +74,86 @@ TEST(TestRmwCheck, test_failing_configuration) {
   EXPECT_TRUE(rcutils_set_env(RMW_IMPLEMENTATION_ENV_VAR_NAME, expected_rmw_impl_env));
   EXPECT_TRUE(rcutils_set_env(RCL_ASSERT_RMW_ID_MATCHES_ENV_VAR_NAME, expected_rmw_id_matches));
 }
+
+// Define dummy comparison operators for rcutils_allocator_t type for use with the Mimick Library
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, >)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, !=)
+
+// Mock internal calls to external libraries to fail
+TEST(TestRmwCheck, test_mock_rmw_impl_check) {
+  {
+    // Fail reading RMW_IMPLEMENTATION_ENV_VAR_NAME
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rcl", rcutils_get_env, "invalid arg");
+    EXPECT_EQ(RCL_RET_ERROR, rcl_rmw_implementation_identifier_check());
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+  {
+    // Fail copying RMW_IMPLEMENTATION_ENV_VAR_NAME env result
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rcl", rcutils_strdup, static_cast<char *>(NULL));
+    EXPECT_EQ(RCL_RET_BAD_ALLOC, rcl_rmw_implementation_identifier_check());
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+  {
+    // Fail reading RCL_ASSERT_RMW_ID_MATCHES_ENV_VAR_NAME
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rcutils_get_env, [](auto, const char ** env_value) {
+        static int counter = 1;
+        *env_value = "";
+        if (counter == 1) {
+          counter++;
+          return static_cast<const char *>(NULL);
+        } else {
+          return "argument env_value is null";
+        }
+      });
+    EXPECT_EQ(RCL_RET_ERROR, rcl_rmw_implementation_identifier_check());
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+  {
+    // Fail copying RCL_ASSERT_RMW_ID_MATCHES_ENV_VAR_NAME env result
+    // Set the variable, as is not set by default
+    const char * expected_rmw_id_matches = NULL;
+    const char * get_env_id_matches_name = rcutils_get_env(
+      RCL_ASSERT_RMW_ID_MATCHES_ENV_VAR_NAME,
+      &expected_rmw_id_matches);
+    EXPECT_FALSE(get_env_id_matches_name);
+    EXPECT_TRUE(rcutils_set_env(RCL_ASSERT_RMW_ID_MATCHES_ENV_VAR_NAME, "some_random_name"));
+
+    // Return an allocated string as strdup would do for the first case
+    // calling the function
+    rcl_allocator_t allocator = rcl_get_default_allocator();
+    size_t dummy_str_size = 3u;
+    char * new_string = static_cast<char *>(allocator.allocate(dummy_str_size, allocator.state));
+    memset(new_string, '\0', dummy_str_size);
+
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rcutils_strdup, [&new_string](auto...) {
+        static int counter = 1;
```

Comment:
@Blast545 FYI if you capture `rcutils_strdup` by value, you can forward calls to it later on.

Reply:
How would I call the function on the internal string?

Reply:
Not sure I follow. But just for the record, you can do:

```c++
auto mock = mocking_utils::patch(
      "lib:rcl", rcutils_strdup, [base = rcutils_strdup](const char * str, auto allocator) {
           return base(str, allocator);
      });
```
where `base` is a dummy name.

Reply:
Cool, I replaced it c4c9fb9